### PR TITLE
feat: do not fail on other nodes when the RPC succeeds on the first node

### PR DESCRIPTION
**Pull request by zhongwencool**
Fixes <issue-or-jira-number>
When updating the configuration, if it has succeeded on the first node, the configuration must be saved successfully on the other nodes even if it has failed. So for both pre_config_update and post_config_update callbacks,  need to support check where they are from, whether it's initiate or replicate.

PS: Previously we only handled post_config_update failed, but now we handle it in pre_config_update as well.

close https://github.com/emqx/emqx/pull/13303

Release version: v/e5.7.2

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx/src/emqx_config_handler.erl [hunk 1]
```diff
@@ -811,3 +893,11 @@ load_prev_handlers() ->
...
 save_handlers(Handlers) ->
     application:set_env(emqx, ?MODULE, Handlers).
+
+get_function_arity(_Module, _Callback, []) ->
+    false;
+get_function_arity(Module, Callback, [Arity | Opts]) ->
+    case erlang:function_exported(Module, Callback, Arity) of
+        true -> Arity;
+        false -> get_function_arity(Module, Callback, Opts)
+    end.
```
Comment (zmstone) [inline apps/emqx/src/emqx_config_handler.erl:903]:
```suggestion
get_function_arity(_Module, _Callback, []) ->
    false;
get_function_arity(Module, Callback, [Arity | More]) ->
    Module = Module:module_info(module), %% ensure module is loaded
    case erlang:function_exported(Module, Callback, Arity) of
        true -> Arity;
        false -> get_function_arity(Module, Callback, More)
    end.
```

File: apps/emqx/src/emqx.erl [hunk 2]
```diff
@@ -195,38 +198,52 @@ get_raw_config(KeyPath, Default) ->
...
+        {remove, Opts},
+        ClusterRpcOpts
     ).
 
 -spec reset_config(emqx_utils_maps:config_key_path(), emqx_config:update_opts()) ->
     {ok, emqx_config:update_result()} | {error, emqx_config:update_error()}.
 reset_config([RootName | SubKeys] = KeyPath, Opts) ->
+    reset_config([RootName | SubKeys] = KeyPath, Opts, #{}).
+
+reset_config([RootName | SubKeys] = KeyPath, Opts, ClusterRpcOpts) ->
```
Comment (zmstone) [inline apps/emqx/src/emqx.erl:246]: add spec, give ClueterPrcOpts a type spec.
-spec cluster_rpc_opts() :: #{kind => ?KIND_INITIATE | ?KIND_REPLICATE}.

File: apps/emqx_resource/src/emqx_resource.erl [hunk 3]
```diff
@@ -50,6 +50,7 @@
     remove_local/1,
     reset_metrics/1,
     reset_metrics_local/1,
+    reset_metrics_local/2,
```
Comment (zmstone) [inline apps/emqx_resource/src/emqx_resource.erl:53]: question: why this change ?

File: apps/emqx_resource/src/emqx_resource.erl [hunk 4]
```diff
@@ -337,8 +338,13 @@ remove_local(ResId) ->
             ok
     end.
 
+%% Tip: Don't delete reset_metrics_local/1, use before v571 rpc
```
Comment (zmstone) [inline apps/emqx_resource/src/emqx_resource.erl:341]:
```suggestion
%% Tip: Don't delete reset_metrics_local/1, use before v572 rpc
```

File: apps/emqx_resource/src/emqx_resource.erl (hunk 3 above)
Comment (zmstone) [reply apps/emqx_resource/src/emqx_resource.erl:53]: I can only find apps/emqx_resource/src/proto/emqx_resource_proto_v2.erl which calls the /1 function.

**Review by zmstone — DISMISSED** [review]

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-2887)
